File: java/src/main/java/org/sparky/model/builders/TableBuilder.java
package org.sparky.model.builders;

import org.sparky.model.Filter;
import org.sparky.model.Rule;
import org.sparky.model.Table;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Created by lloyd on 19/10/2015.
 */
public class TableBuilder {

    private List<FilterBuilder> filters = new ArrayList<>();
    private List<String> headers  = new ArrayList<>();
    private List<RowBuilder> rows = new ArrayList<>();

    public TableBuilder withHeader(String header){
        headers.add(header);
        return this;
    }

    public TableBuilder withRow(RowBuilder row){
        rows.add(row);
        return this;
    }

    public TableBuilder withFilter(FilterBuilder filter){
        filters.add(filter);
        return this;
    }

    public Table build() {
        Map<String, Integer> columns = new HashMap<>();
        int i = 0;
        for(String col : headers){
            columns.put(col, i++);
        }

        String[][] data = new String[rows.size()][headers.size()];
        int rowIndex = 0;
        for(RowBuilder builder : rows){
            for(int colIndex = 0; colIndex < headers.size(); colIndex++){
                data[rowIndex][colIndex] = builder.get(colIndex);
            }
            rowIndex++;
        }

        Map<String, Filter> fl = new HashMap<>();
        for(FilterBuilder b : filters){
            fl.put(b.columnName(), new Filter(b.getRule()));
        }

        return new Table(columns, data, fl);
    }
}

File: net/Sparky/Model/Builder/TableBuilder.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Sparky.Model.Builder
{
    public class TableBuilder
    {
        private readonly List<FilterBuilder> _filters = new List<FilterBuilder>();
        private readonly List<string> _headers = new List<string>();
        private readonly List<RowBuilder> _rows = new List<RowBuilder>();

        public TableBuilder WithHeader(string header)
        {
            _headers.Add(header);
            return this;
        }

        public TableBuilder WithFilter(FilterBuilder b)
        {
            _filters.Add(b);
            return this;
        }

        public TableBuilder WithRow(RowBuilder r)
        {
            _rows.Add(r);
            return this;
        }

        public Table Build()
        {
            var columns = new Dictionary<string, int>();
            int i = 0;
            foreach(var s in _headers)
            {
                columns.Add(s, i++);
            }

            var data = new string[_rows.Count][];
            int rI = 0;
            foreach(var rb in _rows)
            {
                data[rI] = new string[_headers.Count];
                for (var cI = 0; cI < _headers.Count; cI++)
                {
                    data[rI][cI] = rb.GetByIndex(cI);
                }
                rI++;
            }

            var fl = new Dictionary<string, Filter>();
            foreach(var fb in _filters)
            {
                fl.Add(fb.ColumnName(), new Filter(fb.GetRule()));
            }
            return new Table(columns, data, fl);
        }
    }
}
